### PR TITLE
Chown for folder permissions

### PR DESCRIPTION
**Pull request by florianziemen**
the pure mkdir() ignores our requests for the sticky bit.

```
from pathlib import Path
import os
def genp( end) : 
    return Path(f"/scratch/k/k202134/test_pathlib-{end}/")
pts = dict (pure = genp("pure"),
             perm = genp("perm"),
             umask = genp("umask"),
             chown = genp("chown"),
            )
for p in pts.values():
    try: 
        p.rmdir()
    except:
        print ('cannot rmdir ', p)

        
pts['pure'].mkdir()

pts['perm'].mkdir(mode=0o3775)

oldmask = os.umask(0)
pts['umask'].mkdir(mode=0o3775)
os.umask(oldmask)

def mkdirs(file_permissions, path):
        rp = os.path.realpath(path)
        if os.access(rp, os.F_OK):
            print("already there")
            return
        components = Path(rp).parts[1:]
        for i in range (len(components)):
            subpath = Path('/', *components[:i+1])
            if not os.access(subpath, os.R_OK):
                os.mkdir(subpath)
                os.chmod(subpath, file_permissions)

mkdirs(0o3775, pts['chown'])
!ls -dl {genp ('*')}
```
yields
```
drwxrwsr-t 2 k202134 k20200 4096 Feb  7 10:18 /scratch/k/k202134/test_pathlib-chown
drwxr-xr-t 2 k202134 k20200 4096 Feb  7 10:18 /scratch/k/k202134/test_pathlib-perm
drwxr-xr-x 2 k202134 k20200 4096 Feb  7 10:18 /scratch/k/k202134/test_pathlib-pure
drwxrwxr-t 2 k202134 k20200 4096 Feb  7 10:18 /scratch/k/k202134/test_pathlib-umask
```

### CODE REVIEW

**Comment by florianziemen** [comment]
(commits can / should be squashed)

File: slkspec/core.py [hunk 1]
```diff
@@ -151,6 +151,18 @@ def _retrieve_items(self, retrieve_files: list[tuple[str, str]]) -> None:
             for out_file in map(Path, inp_files):
                 (output_dir / out_file.name).chmod(self.file_permissions)
 
+    def mkdirs(self, path):
+        rp = os.path.realpath(path)
+        if os.access(rp, os.F_OK):
```
Comment (observingClouds) [inline slkspec/core.py:156]: From the [python documentation](https://docs.python.org/3/library/os.html#os.access) it seems that `os.access` is not a very good function to use. Can we use an alternative?

File: slkspec/core.py [hunk 2]
```diff
@@ -151,6 +151,18 @@ def _retrieve_items(self, retrieve_files: list[tuple[str, str]]) -> None:
...
+            return
+        components = Path(rp).parts[1:]
+        for i in range (len(components)):
+            subpath = Path("/", *components[:i+1])
+            if not os.access(subpath, os.F_OK):
```
Comment (observingClouds) [inline slkspec/core.py:161]: Shouldn't this rather be `os.W_OK`? Or even better rewritten to a try/except statement? Currently, a folder might exists, but `os.mkdir` still fail because the user doesn't have write permissions.

**Review by observingClouds — COMMENTED** [review]
Thanks for finding this edge-case. I made a few comments, but don't invest too much time into this, because we might be able to get rid of the entire loop over the directories with the new release of pyslk that handles the creation of subfolders itself.

File: slkspec/core.py (hunk 1 above)
Comment (florianziemen) [reply slkspec/core.py:156]: My feeling is that these problems concern problems that we don't really need to care about.
This is really just for checking if the directory is there, or we need to create it.

Comment (florianziemen) [reply slkspec/core.py:156]: We might want to explicitly check for write access though, and throw an exception if the directory exists, but we can't write.

File: slkspec/core.py (hunk 2 above)
Comment (florianziemen) [reply slkspec/core.py:161]: No, we create a subdir, if it does not exist. 
We don't try to create /scratch/ or so, and we don't have write permissions there anyways.

We could again add checks for write permissions starting at the point where we need to start creating (parts of) the path.

Basically the issue here is finding out where in /path/to/data/in/a/subdir/ we have to start creating and adjusting permissions, and the easiest way, I could come up with for doing that was to go bit by bit, check if the subdir is there and create and chmod if necessary.

Comment (florianziemen) [reply slkspec/core.py:161]: The issue of us needing weird subdirectory permissions actually remains independent of the pyslk version...
Group writable sticky stuff is not totally standard, but very helpful for this game with multiple users.
So we might actually have to start putting more rather than less thought into this.

**Comment by observingClouds** [comment]
Hi @florianziemen, could you update your branch with the main branch? Thx.

File: slkspec/core.py (hunk 2 above)
Comment (observingClouds) [reply slkspec/core.py:161]: Well, I would like to keep `slkspec` as general as possible and also useful for other slk tape archives. In the best case only small changes need to be made to use `slkspec` with other tape archive that just offer any way to retrieve files. Introducing handling of the `sticky bit` is very specific and I'd prefer to implement this into the DKRZ-SLK version and not in the last step. Retrieving a file with `slkspec` and `slk retrieve` should result in the same result.

Comment (observingClouds) [reply slkspec/core.py:161]: Could you maybe raise this issue upstream, so we could just forward some specific keywords to `pyslk`?

Comment (observingClouds) [reply slkspec/core.py:161]: CC @neumannd

Comment (neumannd) [reply slkspec/core.py:161]: @observingClouds Is this a feature request for `pyslk` or `slk`? This is no request to add the sticky bit to StrongLink but just for `slk`/`pyslk` to set it for retrieved folders?

You mean a sticky bit as defined as follows (source: https://unix.stackexchange.com/a/79401/182655)?

> The restricted deletion flag or sticky bit is a single bit, whose interpretation depends on the file type. For directories, it
prevents unprivileged users from removing or renaming a file in the directory unless they own the file or the directory; this is called the restricted deletion flag for the directory, and is commonly found on world-writable directories like /tmp.

Comment (observingClouds) [reply slkspec/core.py:161]: @florianziemen seems to be interested in the sticky bit and I think it would be nice if all retrieval methods ( `slk retrieve`, `pyslk.core.retrieve` and `slkspec` behave in the same way. Because all these tools build on top of each other, it would be great if the implementation would be as much down the stack as possible.

Comment (neumannd) [reply slkspec/core.py:161]: `cp` does not support setting the sticky bit while files are copied. There the user needs to set permissions afterwards. `rsync` offers arguments `--chmod` and `--chown`. We could request something similar. What is the advantage of adding this argument to `slk retrieve` compared to just running `chmod` after `slk retrieve?`

Comment (florianziemen) [reply slkspec/core.py:161]: Moin!

So, the really important part ist the gid bit for folders. (2000 of 3775 or so), because it means that subfolders are created with the same user group as the original folder. So, if I download a file into CACHE/arch/run/subdir, subdir belongs to the "intake group", otherwise it becomes part of my group (DKRZ ANW) and no "normal" users can write there (similarly for MPI-ATM vs MPI-OCE / ...).
The sticky bit is nice and seems fitting for the purpose but is less important to me.

So, basically, what we need to do is figure out which folders slk is going to create, and make those gid-bit-y (or change their groups).

Comment (neumannd) [reply slkspec/core.py:161]: What if this is applied on the top folder:

```
chgrp -R ${group} ${folder}
chmod g+rwxs -R ${folder}
```

This should set `$group` permanently for all newly created sub-folders of `$folder`. At least that's what I did with my folder into which I retrieve data for users:

```
chgrp -R ka1209 ${folder}
chmod g+rwxs -R ${folder}
umask 002 ${folder}
setfacl -Rdm g::rwx  ${folder}
```

All users have read permissions and a specific group has write permissions.

I just checked but it seems that the sticky bit of a folder cannot be inherited by its sub-folder.

Comment (florianziemen) [reply slkspec/core.py:161]: yep, so the gid bit ( 2000 ) is passed down. Ok...
Whether we want the suid bit for folders (definitely not for files) is a bit unclear to me.
If we have it, people can't remove other people's files. Seems reasonable (even just to prevent accidental deletions), but also means, I can't get rid of somebody else's broken file if something got stuck mid-way.
We always have the option of calling for divine intervention, but that should not be the default use case.

The other thing we need is the directories to stay group-writable. For me, the umask will generally prevent this, so we need something that takes care of creating/chowning the directories that slk creates upon downloading (assuming slk respects my umask).
Either (1) we are faster and simply create them before slk starts, or (2) we have to _clean up_ after it. I think, I would prefer option (1) as it is more compatible with retrievals by different users and crashed jobs.

**Comment by neumannd** [comment]
@florianziemen If this change was in your [reactivate-pyslk](https://github.com/florianziemen/slkspec/tree/reactivate-pyslk) branch? If so, we can close this PR because my PR #48 was based on your `reactive-pyslk` branch.
